rust/src/shards/gui/layouts/collapsing_header.rs

```diff
@@ -0,0 +1,183 @@
+      defaultOpen: ParamVar::new(false.into()),
+    }
+  }
+}
+
+impl Shard for CollapsingHeader {
+  fn registerName() -> &'static str
+  where
+    Self: Sized,
+  {
+    cstr!("UI.Collapsing")
+  }
+
+  fn hash() -> u32
+  where
+    Self: Sized,
+  {
+    compile_time_crc32::crc32!("UI.Collapsing-rust-0x20200101")
+  }
+
+  fn name(&mut self) -> &str {
+    "UI.Collapsing"
+  }
+
+  fn inputTypes(&mut self) -> &Types {
+    &ANY_TYPES
+  }
+
+  fn outputTypes(&mut self) -> &Types {
+    &ANY_TYPES
+  }
+
+  fn parameters(&mut self) -> Option<&Parameters> {
+    Some(&COLLAPSING_PARAMETERS)
+  }
+
+  fn setParam(&mut self, index: i32, value: &Var) -> Result<(), &str> {
+    match index {
+      0 => Ok(self.heading.set_param(value)),
+      1 => self.contents.set_param(value),
+      2 => Ok(self.defaultOpen.set_param(value)),
+      _ => Err("Invalid parameter index"),
+    }
+  }
+
+  fn getParam(&mut self, index: i32) -> Var {
+    match index {
+      0 => self.heading.get_param(),
+      1 => self.contents.get_param(),
+      2 => self.defaultOpen.get_param(),
+      _ => Var::default(),
+    }
+  }
+
+  fn requiredVariables(&mut self) -> Option<&ExposedTypes> {
+    self.requiring.clear();
+
+    // Add UI.Parents to the list of required variables
+    let exp_info = ExposedInfo {
+      exposedType: EGUI_UI_SEQ_TYPE,
+      name: self.parents.get_name(),
+      help: cstr!("The parent UI objects.").into(),
+      ..ExposedInfo::default()
+    };
+    self.requiring.push(exp_info);
+
+    Some(&self.requiring)
+  }
+
+  fn hasCompose() -> bool {
+    true
+  }
+
+  fn compose(&mut self, data: &InstanceData) -> Result<Type, &str> {
+    if !self.contents.is_empty() {
+      let outputType = self.contents.compose(&data)?;
+      return Ok(outputType);
+    }
+
+    Ok(data.inputType)
+  }
+
+  fn warmup(&mut self, ctx: &Context) -> Result<(), &str> {
+    self.parents.warmup(ctx);
+    self.heading.warmup(ctx);
+    if !self.contents.is_empty() {
+      self.contents.warmup(ctx)?;
+    }
+    self.defaultOpen.warmup(ctx);
+
+    Ok(())
+  }
+
+  fn cleanup(&mut self) -> Result<(), &str> {
+    self.defaultOpen.cleanup();
+    if !self.contents.is_empty() {
+      self.contents.cleanup();
+    }
+    self.heading.cleanup();
+    self.parents.cleanup();
+
+    Ok(())
+  }
+
+  fn activate(&mut self, context: &Context, input: &Var) -> Result<Var, &str> {
+    if self.contents.is_empty() {
+      return Ok(*input);
+    }
+
+    if let Some(ui) = util::get_current_parent(*self.parents.get())? {
+      let heading: &str = self.heading.get().try_into()?;
+      let defaultOpen: bool = self.defaultOpen.get().try_into()?;
+      let header = egui::CollapsingHeader::new(heading).default_open(defaultOpen);
+      if let Some(output) = header
+        .show(ui, |ui| {
+          util::activate_ui_contents(context, input, ui, &mut self.parents, &mut self.contents)
+        })
+        .body_returned
+      {
+        Ok(output?)
```

Comment:
this should be `*input` no?

Reply:
It's the output of the inner shards so that it can be propagated downwards.

It is mostly explain in the README. The idea is that using or not using containers should be transparent as it is just layouting.

Reply:
Maybe this should be just `*input` for simplicity since it's conditional?